### PR TITLE
[llvm][SelectionDAG] Relax llvm.ptrmask's size check on arm64_32

### LINKED ISSUE #94075: llvm.ptrmask broken on arm64_32: Pointers with different index type are not supported by SDAG
The two checks added in eb86de63d901941fba0a54eb41551456447cd789, one in the Verifier and the other in SelectionDAGBuilder.cpp don't quite agree. The Verifier's respects the DataLayout, but the one in SelectionDAG only inspects the in-register type of a pointer, which on arm64_32 differs from its in-memory size.

```
$ cat reduced.ll 
target datalayout = "e-m:o-p:32:32-i64:64-i128:128-n32:64-S128"
target triple = "arm64_32-apple-watchos11.0.0"

define void @rdar125263567() {
entry:
  %aligned_result.i101.i = call ptr @llvm.ptrmask.p0.i32(ptr null, i32 0)
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare ptr @llvm.ptrmask.p0.i32(ptr, i32) #0

attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```

```
$ ./bin/llc reduced.ll
Assertion failed: (PtrVT == Mask.getValueType() && "Pointers with different index type are not supported by SDAG"), function visitIntrinsicCall, file SelectionDAGBuilder.cpp, line 7847.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: ./bin/llc reduced.ll
1.      Running pass 'Function Pass Manager' on module 'reduced.ll'.
2.      Running pass 'AArch64 Instruction Selection' on function '@rdar125263567'
 #0 0x0000000105adb594 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/Users/jonathan_roelofs/llvm-upstream/build/bin/llc+0x103667594)
 #1 0x0000000105adbbb8 PrintStackTraceSignalHandler(void*) (/Users/jonathan_roelofs/llvm-upstream/build/bin/llc+0x103667bb8)
 #2 0x0000000105ad97d4 llvm::sys::RunSignalHandlers() (/Users/jonathan_roelofs/llvm-upstream/build/bin/llc+0x1036657d4)
 #3 0x0000000105adccd4 SignalHandler(int) (/Users/jonathan_roelofs/llvm-upstream/build/bin/llc+0x103668cd4)
 #4 0x0000000192e6eb44 (/usr/lib/system/libsystem_platform.dylib+0x180486b44)
 #5 0x0000000192e37bcc (/usr/lib/system/libsystem_pthread.dylib+0x18044fbcc)
 #6 0x0000000192d44900 (/usr/lib/system/libsystem_c.dylib+0x18035c900)
 #7 0x0000000192d43c24 (/usr/lib/system/libsystem_c.dylib+0x18035bc24)
 #8 0x0000000105769048 llvm::SelectionDAGBuilder::visitIntrinsicCall(llvm::CallInst const&, unsigned int) (/Users/jonathan_roelofs/llvm-upstream/build/bin/llc+0x1032f5048)
 #9 0x000000010574312c llvm::SelectionDAGBuilder::visitCall(llvm::CallInst const&) (/Users/jonathan_roelofs/llvm-upstream/build/bin/llc+0x1032cf12c)
<snip>
```

### PR DESCRIPTION
**Pull request by jroelofs**
Since pointers in memory, as well as the index type are both 32 bits, but in registers pointers are 64 bits, the mask generated by llvm.ptrmask needs to be zero-extended.

Fixes: #94075
Fixes: rdar://125263567

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp [hunk 1]
```diff
@@ -7839,12 +7839,26 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
     return;
   }
   case Intrinsic::ptrmask: {
+    unsigned PtrBits =
+        DAG.getDataLayout().getIndexTypeSizeInBits(I.getOperand(0)->getType());
```
Comment (nikic) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:7843]: This should be checking the pointer size, not the index size. The implementation below is only correct as long as these are the same. (The index size and the mask size are known to be the same by verifier enforcement.)

File: llvm/test/CodeGen/AArch64/lower-ptrmask-arm64_32.ll [hunk 2]
```diff
@@ -0,0 +1,13 @@
+; RUN: llc -mtriple=arm64_32-apple-watchos -stop-after=finalize-isel %s -o - | FileCheck %s
```
Comment (nikic) [inline llvm/test/CodeGen/AArch64/lower-ptrmask-arm64_32.ll:1]: Drop the `-stop-after=finalize-isel` and use update_llc_test_checks.

Comment (jroelofs) [reply llvm/test/CodeGen/AArch64/lower-ptrmask-arm64_32.ll:1]: Not sure why, but UTC doesn't produce meaningful CHECK lines on this one. Am I holding it wrong?

```
$ cat /Users/jonathan_roelofs/llvm-upstream/llvm/test/CodeGen/AArch64/lower-ptrmask-arm64_32.ll
; RUN: llc -mtriple=arm64_32-apple-watchos %s -o - | FileCheck %s

define ptr @issue94075(ptr %p) {
entry:
  %rdar125263567 = call ptr @llvm.ptrmask.p0.i32(ptr %p, i32 4294967288)
  ret ptr %rdar125263567
}
$ ../llvm/utils/update_llc_test_checks.py --llc-binary=./bin/llc /Users/jonathan_roelofs/llvm-upstream/llvm/test/CodeGen/AArch64/lower-ptrmask-arm64_32.ll -v
Scanning for RUN lines in test file: /Users/jonathan_roelofs/llvm-upstream/llvm/test/CodeGen/AArch64/lower-ptrmask-arm64_32.ll
Found 1 RUN lines in /Users/jonathan_roelofs/llvm-upstream/llvm/test/CodeGen/AArch64/lower-ptrmask-arm64_32.ll:
  RUN: llc -mtriple=arm64_32-apple-watchos %s -o - | FileCheck %s
Extracted LLC cmd: llc -mtriple=arm64_32-apple-watchos  -o -
Extracted FileCheck prefixes: ['CHECK']
Rewriting FileCheck prefixes: {'CHECK'}
Writing 10 lines to /Users/jonathan_roelofs/llvm-upstream/llvm/test/CodeGen/AArch64/lower-ptrmask-arm64_32.ll...
$ cat /Users/jonathan_roelofs/llvm-upstream/llvm/test/CodeGen/AArch64/lower-ptrmask-arm64_32.ll
; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
; RUN: llc -mtriple=arm64_32-apple-watchos %s -o - | FileCheck %s

define ptr @issue94075(ptr %p) {
entry:
  %rdar125263567 = call ptr @llvm.ptrmask.p0.i32(ptr %p, i32 4294967288)
  ret ptr %rdar125263567
}
;; NOTE: These prefixes are unused and the list is autogenerated. Do not add tests below this line:
; CHECK: {{.*}}
```

Comment (nikic) [reply llvm/test/CodeGen/AArch64/lower-ptrmask-arm64_32.ll:1]: Probably it doesn't support the arm64_32 triple right now.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/test/CodeGen/AArch64/lower-ptrmask-arm64_32.ll (hunk 2 above)
Comment (jroelofs) [reply llvm/test/CodeGen/AArch64/lower-ptrmask-arm64_32.ll:1]: ohh:
```
def get_run_handler(triple):
    target_handlers = {
        "i686": (scrub_asm_x86, ASM_FUNCTION_X86_RE),
        "x86": (scrub_asm_x86, ASM_FUNCTION_X86_RE),
        "i386": (scrub_asm_x86, ASM_FUNCTION_X86_RE),
        "arm64_32-apple-ios": (scrub_asm_arm_eabi, ASM_FUNCTION_AARCH64_DARWIN_RE),
        "arm64_32-apple-watchos2.0.0": (
            scrub_asm_arm_eabi,
            ASM_FUNCTION_AARCH64_DARWIN_RE,
        ),
 ```

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp [hunk 3]
```diff
@@ -7842,9 +7842,19 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
...
+    // match the index type, but the pointer is 64 bits, so the the mask must be
+    // zero-extended up to 64 bits to match the pointer.
+    EVT PtrVT =
+        TLI.getValueType(DAG.getDataLayout(), I.getOperand(0)->getType());
+    EVT MemVT =
+        TLI.getMemValueType(DAG.getDataLayout(), I.getOperand(0)->getType());
+    assert(PtrVT == Ptr.getValueType());
+    assert(MemVT == Mask.getValueType());
+    if (MemVT != PtrVT)
+      Mask = DAG.getPtrExtOrTrunc(Mask, sdl, PtrVT);
```
[the remark this answers is not in the document]
Comment (nikic) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:7856]: Nevermind, ignore me.

**Pull request merged**